src/panels/ActionsPanel/ActionsPanel.tsx

```diff
@@ -99,52 +167,117 @@ export default function ActionsPanel(): JSX.Element {
+          <button
+            className={classNames(
+              "p-button--positive actions-panel__run-action"
+            )}
+            disabled={disableSubmit}
+            onClick={executeAction}
+          >
+            Run action
+          </button>
         </div>
       </div>
     </Aside>
   );
 }
 
-function generateActionlist(
-  actionData: ActionData | undefined,
-  fetchingActionData: boolean,
-  selectedAction: string | undefined,
-  setSelectedAction: SetSelectedAction
+function onValuesChange(
+  actionName: string,
+  values: ActionOptionValue,
+  optionValues: MutableRefObject<ActionOptionValues>
 ) {
-  if (!actionData && fetchingActionData) return null;
-  if (!actionData && !fetchingActionData)
-    return <div>This charm has not provided any actions</div>;
-  if (!actionData) return null;
-
-  return Object.keys(actionData).map((actionName) => {
-    const action = actionData[actionName];
-    const options: ActionOptions = [];
-
-    Object.keys(action.params.properties).forEach((name) => {
-      const property = action.params.properties[name];
-      options.push({
-        name: name,
-        description: property.description,
-        type: property.type,
-        required: action.params.required.includes(name),
-      });
-    });
-
-    return (
-      <RadioInputBox
-        name={actionName}
-        description={action.description}
-        options={options}
-        onSelect={setSelectedAction}
-        selectedAction={selectedAction}
-        key={actionName}
-      />
-    );
+  const updatedValues: ActionOptionValue = {};
+  Object.keys(values).forEach((key) => {
+    updatedValues[key.replace(`${actionName}-`, "")] = values[key];
   });
+
+  optionValues.current = {
+    ...optionValues.current,
+    [actionName]: updatedValues,
+  };
+}
+
+function enableSubmit(
+  selectedAction: string | undefined,
+  actionData: ActionData,
+  optionsValues: MutableRefObject<ActionOptionValues>,
+  setDisableSubmit: (disable: boolean) => void
+) {
+  if (selectedAction) {
+    if (hasNoOptions(selectedAction, optionsValues.current)) {
+      setDisableSubmit(false);
+      return;
+    }
+    if (
+      requiredPopulated(selectedAction, actionData, optionsValues.current) &&
+      optionsValidate(selectedAction, optionsValues.current)
+    ) {
+      setDisableSubmit(false);
+      return;
+    }
+  }
+  setDisableSubmit(true);
 }
+
+type ValidationFnProps = (
+  selectedAction: string,
+  optionValues: ActionOptionValues
+) => boolean;
+
+type RequiredPopulated = (
+  selectedAction: string,
+  actionData: ActionData,
+  optionValues: ActionOptionValues
+) => boolean;
+
+const hasNoOptions: ValidationFnProps = (selected, optionValues) => {
+  // If there are no options stored then it doesn't have any.
+  if (!optionValues[selected]) {
+    return true;
+  }
+  return Object.keys(optionValues[selected]).length === 0;
+};
+
+const requiredPopulated: RequiredPopulated = (
```

Comment:
You could generate a validation schema for the form to handle this, then the button could check if the form validates before enabling the button: https://formik.org/docs/guides/validation#validationschema.

Reply:
Good call, I actually already have a follow-up task to add proper validation to these fields so I'll address this there.